spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/procedures/CreateChangelogViewProcedure.java

```diff
@@ -183,21 +185,26 @@ private boolean shouldComputeUpdateImages(ProcedureInput input) {
   }
 
   private Dataset<Row> removeCarryoverRows(Dataset<Row> df, boolean netChanges) {
-    Predicate<String> columnsToKeep;
-    if (netChanges) {
-      Set<String> metadataColumn =
-          Sets.newHashSet(
-              MetadataColumns.CHANGE_TYPE.name(),
-              MetadataColumns.CHANGE_ORDINAL.name(),
-              MetadataColumns.COMMIT_SNAPSHOT_ID.name());
-
-      columnsToKeep = column -> !metadataColumn.contains(column);
-    } else {
-      columnsToKeep = column -> !column.equals(MetadataColumns.CHANGE_TYPE.name());
-    }
+    Set<String> metadataColumn =
+        netChanges
+            ? Sets.newHashSet(
+                MetadataColumns.CHANGE_TYPE.name(),
+                MetadataColumns.CHANGE_ORDINAL.name(),
+                MetadataColumns.COMMIT_SNAPSHOT_ID.name())
+            : Sets.newHashSet(MetadataColumns.CHANGE_TYPE.name());
+
+    Predicate<StructField> columnsToDiscard =
+        field ->
+            metadataColumn.contains(field.name())
+                // avoid sort on incomparable columns
+                || field.dataType() instanceof MapType
+                || field.dataType() instanceof BinaryType;
 
     Column[] repartitionSpec =
-        Arrays.stream(df.columns()).filter(columnsToKeep).map(df::col).toArray(Column[]::new);
+        Arrays.stream(df.schema().fields())
+            .filter(Predicate.not(columnsToDiscard))
```

Comment:
Looks like we use it in the repartitioning spec as well, do we need to remove from the columns to be repartitioned as well?

Reply:
I think spark does not seem to complain about have those columns in repartitioned spec, if you have more spark background can you share a bit more what's the best practise and whether we shall remove?

Reply:
I meant to say,
we are now removing the columns from repartitioning spec as well, which was not the case earlier. Is that intended?

Reply:
This breaks the [assumption](https://github.com/apache/iceberg/blob/79620e198009fa243c278c66fd442d107b46206a/spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/RemoveCarryoverIterator.java#L28-L28) of `RemoveCarryoverIterator`.  For example, this is an order can happen now, but it will break `RemoveCarryoverIterator`, as we didn't remove the carryover rows, row1 and row3.
```
row1: (id=1, map=('a', 'a'), op='DELETE')
row2: (id=1, map=('a', 'b')', op='INSERT')
row3: (id=1, map=('a', 'a'), op='INSERT')
```

Reply:
Agreed with @karuppayya that if Spark don't complain the repartition spec with map columns, we should keep them. We can skip them in `sortSpec`. However, that's not enough, we will need to make changes in `ChangeIterator` to make sure skipping-sorting-column works. WDYT?